oauthenticator/gitlab.py

```diff
@@ -144,7 +144,7 @@ def authenticate(self, handler, data=None):
             is_group_specified = True
             user_in_group = yield self._check_group_whitelist(user_id, access_token)
 
-        if self.gitlab_project_id_whitelist:
+        if self.gitlab_project_id_whitelist and not user_in_group:
```

Comment:
maybe add a comment saying that the second check is just to avoid the unecessary API call (just so it's 100% clear that it's an optimization and not part of the semantics of what the check)

Reply:
Added a comment.